File: server/helpers/createGraph.js
const neo4j = require('neo4j-driver');
const News = require('../schema/schema');
const cron = require('node-cron');

// Set up Neo4j driver and session
const driver = neo4j.driver('bolt://localhost:7687', neo4j.auth.basic('neo4j', 'test_password'));
const session = driver.session();

// Simple userId for now (you can change this later to dynamically fetch user info)
const userId = 'user123'; 

async function createGraphFromData() {
    try {
        // Fetch all the news articles from MongoDB
        const newsArticles = await News.find();

        // Iterate over the news articles to create nodes and relationships in Neo4j
        for (const article of newsArticles) {
            // Create a node for each article in Neo4j
            await session.run(
                `MERGE (a:Article {url: $url, createdAt: $createdAt})
                 SET a.timeSpent = $timeSpent`,
                {
                    url: article.url,
                    timeSpent: article.timeSpent,
                    createdAt: article.createdAt
                }
            );

            // Normalize the timeSpent to be between 0 and 1 (you can change the normalization logic)
            const normalizedWeight = article.timeSpent / 100; // Simple normalization (you can adjust this)

            // Create or merge a user node (for now using the static userId)
            await session.run(
                `MERGE (u:User {id: $userId})`,
                { userId: userId }
            );

            // Create a relationship between the user and the article with the normalized weight
            await session.run(
                `MATCH (u:User {id: $userId}), (a:Article {url: $url})
                 MERGE (u)-[:INTERACTED_WITH {weight: $weight}]->(a)`,
                { userId: userId, url: article.url, weight: normalizedWeight }
            );

            // Iterate over tags and create nodes and relationships for each tag
            for (const tag of article.tags) {
                // Merge the tag node (to ensure no duplicates)
                await session.run(
                    `MERGE (t:Tag {name: $tagName})`,
                    { tagName: tag.tag }
                );

                // Create a relationship between the article and the tag
                await session.run(
                    `MATCH (a:Article {url: $url}), (t:Tag {name: $tagName})
                     MERGE (a)-[:HAS_TAG]->(t)`,
                    { url: article.url, tagName: tag.tag }
                );
            }
        }

        console.log('Graph creation completed successfully.');
    } catch (error) {
        console.error('Error creating the graph:', error);
    } finally {
        // Close the Neo4j session
        await session.close();
    }
}

// Set up cron job to run every day at 12:00
// cron.schedule('0 12 * * *', () => {
//     console.log('Running graph creation process at 12:00...');
//     createGraphFromData().then(() => {
//         console.log('Graph creation process is finished.');
//     }).catch(err => {
//         console.error('Error during graph creation process:', err);
//     });
// });

// console.log('Scheduled task is running...');

module.exports = createGraphFromData;

File: server/pythonFiles/create_graph.py
from neo4j import GraphDatabase
import sys
import json

# Neo4j connection setup
URI = "bolt://localhost:7687"  # Update if different
USERNAME = "neo4j"
PASSWORD = "test_password"

class GraphCreator:
    def __init__(self, uri, username, password):
        self.driver = GraphDatabase.driver(uri, auth=(username, password))

    def close(self):
        self.driver.close()

    def create_graph(self, data):
        with self.driver.session() as session:
            for article in data:
                # Create article node
                session.run(
                    """
                    MERGE (a:Article {url: $url})
                    SET a.category = $category, a.createdAt = $createdAt
                    """, 
                    url=article["url"], 
                    category=article["category"], 
                    createdAt=article["createdAt"]
                )

                # Create or merge category node
                session.run(
                    """
                    MERGE (c:Category {name: $category})
                    """, 
                    category=article["category"]
                )

                # Create relationship between article and category
                session.run(
                    """
                    MATCH (a:Article {url: $url})
                    MATCH (c:Category {name: $category})
                    MERGE (a)-[:BELONGS_TO]->(c)
                    """, 
                    url=article["url"], 
                    category=article["category"]
                )

                # Create user -> article relationship
                session.run(
                    """
                    MATCH (a:Article {url: $url})
                    MERGE (u:User {id: $userId})
                    MERGE (u)-[r:READ]->(a)
                    SET r.timeSpent = $timeSpent
                    """, 
                    userId="user789", 
                    url=article["url"], 
                    timeSpent=article["timeSpent"]
                )

                # Create tags and article -> tag relationships
                for tag in article["tags"]:
                    session.run(
                        """
                        MATCH (a:Article {url: $url})
                        MERGE (t:Tag {name: $tagName})
                        MERGE (a)-[r:HAS_TAG]->(t)
                        SET r.weight = $weight
                        """, 
                        url=article["url"], 
                        tagName=tag["tag"], 
                        weight=tag["weight"]
                    )

    def normalize_read_weights(self):
        """
        Normalize the timeSpent weights for the READ relationships dynamically.
        """
        with self.driver.session() as session:
            # Step 1: Fetch the maximum timeSpent value
            result = session.run(
                """
                MATCH (:User)-[r:READ]->(:Article)
                RETURN max(r.timeSpent) AS maxTimeSpent
                """
            )
            max_time_spent = result.single()["maxTimeSpent"]

            if max_time_spent and max_time_spent > 0:
                # Step 2: Normalize all timeSpent weights
                session.run(
                    """
                    MATCH (:User)-[r:READ]->(:Article)
                    SET r.timeSpent = r.timeSpent / $maxTimeSpent
                    """, 
                    maxTimeSpent=max_time_spent
                )
                print(f"Normalization completed. Max timeSpent was: {max_time_spent}")
            else:
                print("Normalization skipped. No valid timeSpent values found.")

if __name__ == "__main__":
    # Read JSON data from stdin
    input_data = json.loads(sys.stdin.read())
    graph_creator = GraphCreator(URI, USERNAME, PASSWORD)
    try:
        # Step 1: Create the graph
        graph_creator.create_graph(input_data)
        print("Graph created successfully")

        # Step 2: Normalize the timeSpent weights
        graph_creator.normalize_read_weights()
    except Exception as e:
        print(f"Error during execution: {e}")
    finally:
        graph_creator.close()
